server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1042,57 +1048,158 @@ public boolean isReadOnly() {
+                public void onFailure(String source, Exception e) {
+                    listener.onFailure(
+                        new RepositoryException(metadata.name(), "Failed to execute cluster state update [" + source + "]", e));
+                }
+
+                @Override
+                public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                    assert assertExpectedGeneration(newState);
+                    setPendingStep.onResponse(newGen);
+                }
+
+                private boolean assertExpectedGeneration(ClusterState newState) {
+                    final RepositoryMetaData repoState = getRepoMetaData(newState);
+                    assert newGen == repoState.pendingGeneration()
+                        : "State [" + repoState + "] did not contain assumed pending generation [" + newGen + "]";
+                    return true;
+                }
+            });
+
+        // Step 2: Write new index-N blob to repository and update index.latest
+        setPendingStep.whenComplete(newGen -> threadPool().executor(ThreadPool.Names.SNAPSHOT).execute(ActionRunnable.wrap(listener, l -> {
             if (latestKnownRepoGen.get() >= newGen) {
                 throw new IllegalArgumentException(
-                    "Tried writing generation [" + newGen + "] but repository is at least at generation [" + newGen + "] already");
+                    "Tried writing generation [" + newGen + "] but repository is at least at generation [" + latestKnownRepoGen.get()
+                        + "] already");
             }
             // write the index file
             final String indexBlob = INDEX_FILE_PREFIX + Long.toString(newGen);
             logger.debug("Repository [{}] writing new index generational blob [{}]", metadata.name(), indexBlob);
             writeAtomic(indexBlob,
                 BytesReference.bytes(repositoryData.snapshotsToXContent(XContentFactory.jsonBuilder(), writeShardGens)), true);
-            final long latestKnownGen = latestKnownRepoGen.updateAndGet(known -> Math.max(known, newGen));
-            if (newGen < latestKnownGen) {
-                // Don't mess up the index.latest blob
-                throw new IllegalStateException(
-                    "Wrote generation [" + newGen + "] but latest known repo gen concurrently changed to [" + latestKnownGen + "]");
-            }
             // write the current generation to the index-latest file
             final BytesReference genBytes;
             try (BytesStreamOutput bStream = new BytesStreamOutput()) {
                 bStream.writeLong(newGen);
                 genBytes = bStream.bytes();
             }
             logger.debug("Repository [{}] updating index.latest with generation [{}]", metadata.name(), newGen);
+
             writeAtomic(INDEX_LATEST_BLOB, genBytes, false);
-            // delete the N-2 index file if it exists, keep the previous one around as a backup
```

Comment:
I wonder if keeping `n-1` around helps with restoring a snapshot from a repo which is concurrently being snapshotted into. (Assume e.g. a scenario where a cluster is taking regular snapshots, and the cluster is being cloned from the snapshot)

Reply:
I don't think so. All the shard generations will be off if you read an outdated `index-N` anyway.  The way the getting of `RepositoryData` works now, with retrying if the generation changes concurrently, being able to successfully read an outdated index-N (and then failing on missing shard generations) instead of retrying for the newest `RepositoryData` actually introduceds a new source of restore failures come to think of it :)